### DISCUSSION TITLE
Strange behavior of prefetch in pages router with or without middleware

### DISCUSSION THREAD

**Discussion by mogulist** [vercel/next.js#66071] [Help]
### Summary

In recent version of next.js, prefetch in pages router does not work as documented.

### Additional information

Prefetch does not occur on pages with links to SSR pages. Adding a very simple middleware causes prefetch to occur, but the response to the prefetch is an empty object. After much trouble, I figured out how to get the prefetch to retrieve the correct values. If the x-middleware-prefetch header is removed in the middleware, prefetch for SSR pages correctly retrieves data.

For pages with links to SSG pages, prefetch works correctly regardless of the presence of middleware.

It is strange that neither the middleware nor prefetch documentation mentions that middleware needs to be added to prefetch links to SSR pages. I'm curious whether this should be considered a bug or if it is an intentional feature.

After searching through discussions, I found a post mentioning that prefetch worked in version 12 but stopped working after upgrading to version 14(https://github.com/vercel/next.js/discussions/64914). This seems to be the same issue I am experiencing.

### Example

https://github.com/mogulist/test-middleware-prefetch-on-pages-router

**Comment by victorcarvalhosp** [comment]
Did you find a workaround for this? I'm having the exact same issue

**Comment by barry166** [comment]
This is expected for Pages Router pages that use `getServerSideProps` when middleware is in the path.

The client marks that request with `x-middleware-prefetch: 1`, and the server intentionally bails for non-SSG pages instead of running SSR during a prefetch. That is why the response is effectively empty / `x-middleware-skip`, and why removing the header makes it fetch data: you are turning the request back into a real data request, not a middleware prefetch.

So the safe workaround is: don't rely on `next/link` prefetch to warm `getServerSideProps` data behind middleware. Let the navigation fetch it, move that data to SSG/ISR where possible, or prefetch the underlying API/client data yourself. Stripping `x-middleware-prefetch` in middleware can work as a local hack, but it also opts back into SSR work during prefetch, which is exactly what Next is trying to avoid here.

**Comment by yakubka** [comment]
This is the documented behavior for prefetch with middleware in the Pages Router, though the docs do not make it obvious.

When middleware is present, Next.js intercepts prefetch requests and runs them through middleware first. The middleware runtime adds an x-middleware-prefetch header to signal that the request is a prefetch. By default, middleware rewrites the prefetch response to an empty object for SSR pages because the full page data may depend on the actual request context (cookies, auth state, etc.) that is not yet available at prefetch time. This is intentional so middleware can gate content without leaking unauthed data in prefetch.

The workaround you found, stripping x-middleware-prefetch in middleware, causes Next to fall through to the normal SSR data fetch, which is why prefetch starts returning actual page props.

A cleaner way to do this is to check for the header in middleware and conditionally skip your logic for prefetch requests:

In your middleware file, read the x-middleware-prefetch header with request.headers.get. If it is set, call next() early so the prefetch goes through to the SSR page as a normal fetch rather than being short-circuited.

For SSG pages the issue does not appear because static page data is always safe to return at prefetch time since it has no per-request context.

The reason this changed between version 12 and 14 is that middleware was made more aggressive about intercepting prefetch requests starting in version 13 as part of the App Router rollout. The pages router inherits the same behavior.
